Content.Goobstation.Server/HisGrace/HisGraceSystem.cs

```diff
@@ -0,0 +1,469 @@
+    {
+        if (hisGrace.CurrentState is HisGraceState.Dormant or HisGraceState.Death or HisGraceState.Ascended)
+            return;
+
+        if (TerminatingOrDeleted(uid) || hisGrace.User is not { } user)
+        {
+            hisGrace.CurrentState = HisGraceState.Dormant;
+            return;
+        }
+
+        HandleUserDistance(uid, hisGrace, user);
+        HandleGroundAttacks(uid, hisGrace, melee, xform);
+        UpdateHungerState(hisGrace);
+        ProcessHungerTick(hisGrace, user);
+    }
+
+    private void HandleUserDistance(EntityUid uid, HisGraceComponent hisGrace, EntityUid user)
+    {
+        if (_timing.CurTime <= hisGrace.NextUserAttack
+        || _lookup.GetEntitiesInRange(uid, 1f).Contains(user))
+            return;
+
+        var popUp = Loc.GetString("hisgrace-too-far");
+        _popup.PopupEntity(popUp, user, user, PopupType.LargeCaution);
+
+        _damageable.TryChangeDamage(user, hisGrace.BaseDamage, targetPart: TargetBodyPart.Torso, ignoreResistances: true);
+
+        hisGrace.NextUserAttack = _timing.CurTime + hisGrace.TickDelay;
+    }
+
+    private void HandleGroundAttacks(EntityUid uid, HisGraceComponent hisGrace, MeleeWeaponComponent melee, TransformComponent xform)
+    {
+        if (hisGrace.IsHeld && hisGrace.Holder == hisGrace.User
+        || _timing.CurTime < hisGrace.NextGroundAttack)
+            return;
+
+        var nearbyEnts = _lookup.GetEntitiesInRange(uid, 1f);
+
+        // dont attack if the entity is the user, and dont if the entity is in a container (e.g, already devoured)
+        foreach (var entity in nearbyEnts.Where(entity => HasComp<MobStateComponent>(entity) // malicious foreach loop
+        && entity != hisGrace.User
+        && !_containerSystem.IsEntityOrParentInContainer(entity)))
+        {
+            /// get co-ordinates for animation
+            var coordinates = _transform.GetMapCoordinates(uid);
+            var angle = _transform.GetRelativePosition(xform, entity, GetEntityQuery<TransformComponent>()).ToAngle();
+
+            // do damage and animation
+            _damageable.TryChangeDamage(entity, melee.Damage, targetPart: TargetBodyPart.Torso, origin: uid);
+            _melee.DoLunge(uid, uid, angle, coordinates.Position, null, angle, false, false);
+
+            _audio.PlayPvs(melee.HitSound, uid);
+            _popup.PopupEntity(Loc.GetString("hisgrace-attack-popup", ("target", Name(entity))), uid, PopupType.LargeCaution);
+
+            TryDevour(hisGrace, entity);
+
+            hisGrace.NextGroundAttack = _timing.CurTime + hisGrace.TickDelay;
+            break;
+        }
+    }
+
+    private void UpdateHungerState(HisGraceComponent hisGrace)
+    {
+        // if the current hunger is less than the current threshold min, a downgrade is needed.
+        var downgradeNeeded =
+        hisGrace.StateThresholds.TryGetValue(hisGrace.CurrentState, out var currentThreshold)
+        && hisGrace.Hunger < currentThreshold.Threshold;
```

Comment:
I kinda hate this but the variables are too long.
```suggestion
        var downgradeNeeded = hisGrace.StateThresholds.TryGetValue(hisGrace.CurrentState, out var currentThreshold)
            && hisGrace.Hunger < currentThreshold.Threshold;